src/main/java/io/wispforest/lavender/structure/StructureTemplate.java

```diff
@@ -163,224 +287,330 @@ public int countValidStates(World world, BlockPos anchor, BlockRotation rotation
-                } else if (xSize != rowElement.getAsString().length()) {
-                    throw new JsonParseException("Every row must have the same length");
+        @NotNull
+        private final BlockState[] previewStates;
+
+        public TagBlockStatePredicate(@NotNull RegistryEntryList.Named<Block> tagEntries, @NotNull Map<String, String> properties) {
+            this.vagueProperties = properties;
+            this.tag = tagEntries.getTag();
+            this.previewStates = tagEntries.stream().map(entry -> {
+                var block = entry.value();
+                var state = block.getDefaultState();
+
+                for (var propAndValue : this.vagueProperties.entrySet()) {
+                    Property prop = block.getStateManager().getProperty(propAndValue.getKey());
+                    if (prop == null) continue;
+
+                    Optional<Comparable> value = prop.parse(propAndValue.getValue());
+                    if (value.isEmpty()) continue;
+
+                    state = state.with(prop, value.get());
                 }
-            }
+
+                return state;
+            }).toArray(BlockState[]::new);
         }
 
-        var result = new BlockStatePredicate[xSize][][];
-        for (int x = 0; x < xSize; x++) {
-            result[x] = new BlockStatePredicate[ySize][];
-            for (int y = 0; y < ySize; y++) {
-                result[x][y] = new BlockStatePredicate[zSize];
+        @NotNull
+        @Override
+        public Result test(@NotNull BlockState state) {
+            if (!state.isIn(this.tag))
+                return Result.NO_MATCH;
+
+            for (var propAndValue : this.vagueProperties.entrySet()) {
+                var prop = state.getBlock().getStateManager().getProperty(propAndValue.getKey());
+                if (prop == null)
+                    return Result.BLOCK_MATCH;
+
+                var expected = prop.parse(propAndValue.getValue());
+                if (expected.isEmpty())
+                    return Result.BLOCK_MATCH;
+
+                if (!state.get(prop).equals(expected.get()))
+                    return Result.BLOCK_MATCH;
             }
+
+            return Result.STATE_MATCH;
         }
 
-        for (int y = 0; y < layersArray.size(); y++) {
-            var layer = (JsonArray) layersArray.get(y);
-            for (int z = 0; z < layer.size(); z++) {
-                var row = layer.get(z).getAsString();
-                for (int x = 0; x < row.length(); x++) {
-                    char key = row.charAt(x);
+        @Override
+        public BlockState[] previewBlockstates() {
+            return this.previewStates;
+        }
+    }
 
-                    BlockStatePredicate predicate;
-                    if (keys.containsKey(key)) {
-                        predicate = keys.get(key);
+    private record StructureTemplateRenderView(@NotNull World world, @NotNull StructureTemplate template) implements BlockRenderView {
```

Comment:
Why is this a record now? It was doing perfectly fine as a local class

Reply:
because intellij was yelling at me saying to make it a record lol

Reply:
> Why is this a record now? It was doing perfectly fine as a local class

I pulled it out as an inner class as I find local classes to be a bit of a codesmell.

Would you like me to revert this?